storage/client/queries.go

```diff
@@ -394,28 +394,50 @@ func (qf QueryFactory) RuntimeBlocksQuery() string {
 
 func (qf QueryFactory) RuntimeTransactionsQuery() string {
 	return fmt.Sprintf(`
-		SELECT 
-				txs.round,
-				txs.tx_index,
-				txs.tx_hash,
-				txs.tx_eth_hash,
-				txs.timestamp,
-				txs.raw,
-				txs.result_raw
-			FROM %[1]s.runtime_transactions AS txs
-			LEFT JOIN %[1]s.runtime_related_transactions AS rel_accounts ON txs.round = rel_accounts.tx_round 
-				AND txs.tx_index = rel_accounts.tx_index
-				AND txs.runtime = rel_accounts.runtime
-				-- When related_address ($3) is NULL and hence we do no filtering on it, avoid the join altogether.
-				-- Otherwise, every tx will be returned as many times as there are related addresses for it. 
-				AND $3::text IS NOT NULL
-			WHERE (txs.runtime = '%[2]s') AND
-						($1::bigint IS NULL OR txs.round = $1::bigint) AND
-						($2::text IS NULL OR txs.tx_hash = $2::text OR txs.tx_eth_hash = $2::text) AND
-						($3::text IS NULL OR rel_accounts.account_address = $3::text)
-		ORDER BY txs.round DESC, txs.tx_index DESC
-		LIMIT $4::bigint
-		OFFSET $5::bigint`, qf.chainID, qf.runtime)
+		SELECT
+			txs.round,
+			txs.tx_index,
+			txs.tx_hash,
+			txs.tx_eth_hash,
+			txs.timestamp,
+			txs.raw,
+			txs.result_raw,
+			json_object_agg(pre.address, pre.address_data) AS preimages
+		FROM %[1]s.runtime_transactions AS txs
+		JOIN %[1]s.runtime_related_transactions AS rel ON txs.runtime = rel.runtime
+			AND txs.round = rel.tx_round
+			AND txs.tx_index = rel.tx_index
+		JOIN %[1]s.address_preimages AS pre ON rel.account_address = pre.address
+		WHERE (txs.runtime, txs.round, txs.tx_index) IN
+			(
+				SELECT
+					txs.runtime,
+					txs.round,
+					txs.tx_index
+				FROM %[1]s.runtime_transactions AS txs
+				LEFT JOIN %[1]s.runtime_related_transactions AS rel_accounts ON txs.round = rel_accounts.tx_round
+					AND txs.tx_index = rel_accounts.tx_index
+					AND txs.runtime = rel_accounts.runtime
+					-- When related_address ($3) is NULL and hence we do no filtering on it, avoid the join altogether.
+					-- Otherwise, every tx will be returned as many times as there are related addresses for it.
+					AND $3::text IS NOT NULL
+				WHERE (txs.runtime = '%[2]s') AND
+							($1::bigint IS NULL OR txs.round = $1::bigint) AND
+							($2::text IS NULL OR txs.tx_hash = $2::text OR txs.tx_eth_hash = $2::text) AND
+							($3::text IS NULL OR rel_accounts.account_address = $3::text)
+				ORDER BY txs.round DESC, txs.tx_index DESC
+				LIMIT $4::bigint
+				OFFSET $5::bigint
+			)
+		GROUP BY
+			txs.round,
+			txs.tx_index,
+			txs.tx_hash,
+			txs.tx_eth_hash,
+			txs.timestamp,
+			txs.raw,
+			txs.result_raw
+		`, qf.chainID, qf.runtime)
```

Comment:
I find this more readable when the inner and outer query are inverted:
```sql
SELECT 
  txs.*,  -- all the fields from GROUP BY
  (              
    SELECT json_object_agg(pre.address, pre.address_data)
    FROM oasis_3.runtime_related_transactions rel
    JOIN oasis_3.address_preimages pre ON rel.account_address = pre.address
    WHERE txs.runtime = rel.runtime AND txs.round = rel.tx_round AND txs.tx_index = rel.tx_index
  ) AS eth_addr_lookup
  FROM oasis_3.runtime_transactions txs
  -- LEFT JOIN rel_accounts
  -- WHERE ... 
  LIMIT 1000;
```

It avoids the expansion-and-then-contraction of the number of rows where we first JOIN and then GROUP many of the resulting rows.

I also tested the speed, and it's a little faster (for limit=10 or 100; a little slower for limit=1000, which will probably see less use).

WDYT?

Reply:
Oh yeah. This is much nice than the `GROUP BY`.